src/nodes/hypertable_modify.c

```diff
@@ -93,6 +95,276 @@ get_chunk_dispatch_states(PlanState *substate)
+				leftop = ((RelabelType *) leftop)->arg;
+			if (IsA(rightop, RelabelType))
+				rightop = ((RelabelType *) rightop)->arg;
+
+			if (IsA(leftop, Var) && IsA(rightop, Const))
+			{
+				var = (Var *) leftop;
+				arg_value = (Const *) rightop;
+			}
+			else if (IsA(rightop, Var) && IsA(leftop, Const))
+			{
+				var = (Var *) rightop;
+				arg_value = (Const *) leftop;
+			}
+			else
+				return false;
+
+			column_name = get_attname(context->chunk_rte->relid, var->varattno, false);
+			if (is_segmentby_col(ht_compression_info, column_name))
+			{
+				tce = lookup_type_cache(var->vartype, TYPECACHE_BTREE_OPFAMILY);
+				op_strategy = get_op_opfamily_strategy(opexpr->opno, tce->btree_opf);
+				if (op_strategy == BTLessStrategyNumber || op_strategy == BTEqualStrategyNumber ||
+					op_strategy == BTGreaterStrategyNumber)
+					build_scankeys(context, column_name, arg_value, op_strategy);
+			}
+		}
+		break;
+		case T_NullTest:
+		{
+			NullTest *ntest = (NullTest *) node;
+			if (IsA(ntest->arg, Var))
+			{
+				int attrno = 0;
+				Var *var = (Var *) ntest->arg;
+				column_name = get_attname(context->chunk_rte->relid, var->varattno, false);
+				if (is_segmentby_col(ht_compression_info, column_name) &&
+					is_segmentby_col_null(context, column_name, &attrno))
+				{
+					context->null_columns = bms_add_member(context->null_columns, attrno);
+				}
+			}
+		}
+		break;
+		default:
+			return false;
+	}
+	return false;
+}
+
+static void
+build_scankeys_for_segmentby_qual(PlanState *ps, HypertableModifyState *context)
+{
+	ListCell *lc;
+	context->chunk_rte = rt_fetch(((Scan *) ps->plan)->scanrelid, ps->state->es_range_table);
+	foreach (lc, ps->plan->qual)
+	{
+		parse_predicate(lfirst(lc), context);
+	}
+}
+
+/*
+ * Traverse the plan tree to look for Scan nodes on uncompressed chunks.
+ * Once Scan node is found check if chunk is compressed, if so then save
+ * the chunk in HypertableModifyState to be used during plan execution.
+ * We also parse the WHERE quals to get information about predicates.
+ */
+static bool
+collect_chunks_from_scan(PlanState *ps, HypertableModifyState *context)
+{
+	RangeTblEntry *rte = NULL;
+	Chunk *chunk;
+	if (ps == NULL || ts_guc_enable_transparent_decompression == false)
+		return false;
+
+	switch (nodeTag(ps))
+	{
+		case T_SeqScanState:
+		case T_SampleScanState:
+		case T_IndexScanState:
+		case T_IndexOnlyScanState:
+		case T_BitmapHeapScanState:
+		case T_TidScanState:
+		case T_TidRangeScanState:
+			rte = rt_fetch(((Scan *) ps->plan)->scanrelid, ps->state->es_range_table);
+			chunk = ts_chunk_get_by_relid(rte->relid, false);
+			if (chunk && ts_chunk_is_compressed(chunk))
+			{
+				context->chunks = lappend(context->chunks, chunk);
+				if (ps->plan->qual && !context->chunk_quals_processed)
+				{
+					/* check if perdicate has SEGMENT BY columns, then build scan keys */
+					build_scankeys_for_segmentby_qual(ps, context);
+					context->chunk_quals_processed = true;
+				}
```

Comment:
So we only do this for the first chunk, right? I think scankeys have the attribute numbers inside, and they might not match for different chunks. We have some examples in `compression_alter.sql`, basically you add a chunk, compress it, drop a column from table, add another chunk, compress it. And there you have two compressed chunks with different sequences of columns.

Reply:
I had the same doubt, then i did not have a valid testcase to prove it. Thanks for finding this issue. 
Now for each chunk i do get scankeys.